Neos.Flow/Classes/Mvc/Routing/RouteConfiguration.php

```diff
@@ -0,0 +1,120 @@
+     *
+     * @return Routes
+     */
+    public function getRoutes(): Routes
+    {
+        if ($this->routesCreated === false) {
+            $this->createRoutesFromConfiguration();
+        }
+        return $this->routes;
+    }
+
+    /**
+     * Creates \Neos\Flow\Mvc\Routing\Route objects from the injected routes
+     * configuration.
+     *
+     * @return void
+     * @throws InvalidRouteSetupException
+     */
+    protected function createRoutesFromConfiguration(): void
+    {
+        $routesConfiguration = $this->configurationManager->getConfiguration(ConfigurationManager::CONFIGURATION_TYPE_ROUTES);
+        $routes = [];
+        $routesWithHttpMethodConstraints = [];
+        foreach ($routesConfiguration as $routeConfiguration) {
+            $route = new Route();
+            if (isset($routeConfiguration['name'])) {
+                $route->setName($routeConfiguration['name']);
+            }
+            $uriPattern = $routeConfiguration['uriPattern'];
+            $route->setUriPattern($uriPattern);
+            if (isset($routeConfiguration['defaults'])) {
+                $route->setDefaults($routeConfiguration['defaults']);
+            }
+            if (isset($routeConfiguration['routeParts'])) {
+                $route->setRoutePartsConfiguration($routeConfiguration['routeParts']);
+            }
+            if (isset($routeConfiguration['toLowerCase'])) {
+                $route->setLowerCase($routeConfiguration['toLowerCase']);
+            }
+            if (isset($routeConfiguration['appendExceedingArguments'])) {
+                $route->setAppendExceedingArguments($routeConfiguration['appendExceedingArguments']);
+            }
+            if (isset($routeConfiguration['httpMethods'])) {
+                if (isset($routesWithHttpMethodConstraints[$uriPattern]) && $routesWithHttpMethodConstraints[$uriPattern] === false) {
+                    throw new InvalidRouteSetupException(sprintf('There are multiple routes with the uriPattern "%s" and "httpMethods" option set. Please specify accepted HTTP methods for all of these, or adjust the uriPattern', $uriPattern), 1365678427);
+                }
+                $routesWithHttpMethodConstraints[$uriPattern] = true;
+                $route->setHttpMethods($routeConfiguration['httpMethods']);
+            } else {
+                if (isset($routesWithHttpMethodConstraints[$uriPattern]) && $routesWithHttpMethodConstraints[$uriPattern] === true) {
+                    throw new InvalidRouteSetupException(sprintf('There are multiple routes with the uriPattern "%s" and "httpMethods" option set. Please specify accepted HTTP methods for all of these, or adjust the uriPattern', $uriPattern), 1365678432);
+                }
+                $routesWithHttpMethodConstraints[$uriPattern] = false;
+            }
+            $routes[] = $route;
+        }
+        $this->routes = Routes::fromArray($routes);
+
+        $this->emitRoutesCreated($this->routes);
```

Comment:
I would suggest not to expose the internals and make this class an extension point.
Instead we could add a signal to the `Router` itself and/or introduce a very simple interface like:
```php
interface RouteConfiguration {
    public function getRoutes(): Routes;
}
```

(even though `RoutesProvider` or `RoutesSource` might be a better name)

Reply:
To clarify: The extension point, is to take a package like this

https://github.com/wbsply/route-annotation

and give it a point to hook into the configuration. In package above, I override the RouterInterface implementation to add annotated routes. 

What you would suggest, is to give the extension possibility to the `Router` class? 

@bwaidelich  Can you write a scratch file of your idea and I can adapt the PR :)